src/dotnet/Fable.Compiler/Transforms/MonadicTrampoline.fs

```diff
@@ -1,51 +1,23 @@
-﻿// Source: http://fssnip.net/dK/title/Monadic-Trampoline
-module MonadicTrampoline
+﻿module MonadicTrampoline
 
 #if NETSTANDARD2_0 || !FABLE_COMPILER
-type TrampolineValue<'T> =
-    | DelayValue of Delay<'T>
-    | ReturnValue of Return<'T>
-    | BindValue of IBind<'T>
 
-and ITrampoline<'T> =
-    abstract member Value : TrampolineValue<'T>
-    abstract member Run : unit -> 'T
+type Thunk<'T> =
+    | DelayValue of (unit -> Thunk<'T>)
+    | ReturnValue of 'T
 
-and Delay<'T>(f : unit -> ITrampoline<'T>) =
-    member self.Func = f
-    interface ITrampoline<'T> with
-        member self.Value = DelayValue self
-        member self.Run () = (f ()).Run()
-
-and Return<'T>(x :'T) =
-    member self.Value = x
-    interface ITrampoline<'T> with
-        member self.Value = ReturnValue self
-        member self.Run () = x
-
-and IBind<'T> =
-    abstract Bind<'R> : ('T -> ITrampoline<'R>) -> ITrampoline<'R>
-
-and Bind<'T, 'R>(trampoline : ITrampoline<'T>, f : ('T -> ITrampoline<'R>)) =
-    interface IBind<'R> with
-        member self.Bind<'K>(f' : 'R -> ITrampoline<'K>) : ITrampoline<'K> =
-            new Bind<'T, 'K>(trampoline, fun t -> new Bind<'R, 'K>(f t, (fun r -> f' r)) :> _) :> _
-    interface ITrampoline<'R> with
-        member self.Value = BindValue self
-        member self.Run () =
-            match trampoline.Value with
-            | BindValue b -> b.Bind(f).Run()
-            | ReturnValue r -> (f r.Value).Run()
-            | DelayValue d -> (new Bind<'T, 'R>(d.Func (), f) :> ITrampoline<'R>).Run()
+let rec run = function
+    | DelayValue f -> f () |> run
+    | ReturnValue x -> x
 
 type TrampolineBuilder() =
-    member self.Bind(tr, f) = new Bind<'T, 'R>(tr, f) :> ITrampoline<'R>
-    member self.Delay f = new Delay<_>(f) :> ITrampoline<_>
-    member self.Return a = new Return<_>(a) :> ITrampoline<_>
-    member self.ReturnFrom a = a
+    member __.Bind(thunk, f) = DelayValue (fun () -> run thunk |> f)
```

Comment:
So this is working? I think for some reason it didn't work for me so I had to keep `BindValue` with the boxed parameter. If this implementation is working, we should be able to unify it with the Fable version because the problem was Fable couldn't optimize mutually recursive functions, but now there's only `run`.